url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/c64utils/IPPacket.java

```diff
@@ -0,0 +1,159 @@
+package com.dreamfabric.c64utils;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Describe class IPPacket here.
+ *
+ *
+ * Created: Thu Apr 12 15:30:35 2007
+ *
+ * @author <a href="mailto:Joakim@BOTBOX"></a>
+ * @version 1.0
+ */
+public class IPPacket {
+
+  byte[] data;
+  byte[] header;
+  /**
+   * Creates a new <code>IPPacket</code> instance.
+   *
+   */
+  public IPPacket() {
+  }
+  // Should take a data array and split it into header and data!!!!
+  public IPPacket(byte[] header, byte[] data) {
+    this.header = header;
+    this.data = data;
+  }
+
+  public void readIPPacket(InputStream fs) throws IOException {
+    byte[] header = new byte[20];
+    // Search for next 0x45 (typical IP header size!)
+    int pos = 0;
+    int c = 0;
+    boolean startFound = false;
+    while (pos < 20 && (c = fs.read()) != -1) {
+      if (c == 0x45) {
+	startFound = true;
+      }
+      if (startFound) {
+	header[pos++] = (byte) c;
+      }
+    }
+    this.header = header;
+    int dataLen = getTotalLength() - 20;
+    data = new byte[dataLen];
+    for (int i = 0, n = dataLen; i < n && (c = fs.read()) != -1; i++) {
+      data[i] = (byte) (c & 0xff);
+    }
+  }
+
+  public int getVersion() {
+    return (header[0] >> 4) & 0xf;
+  }
+
+  public int getHeaderLengthBytes() {
+    return 4 * ((header[0]) & 0xf);
+  }
+
+  public int getServiceType() {
+    return header[1] & 0xff;
+  }
+
+  // High byte first?
+  public int getTotalLength() {
+    return get16(2);
+  }
+
+  public int getID() {
+    return get16(4);
+  }
+
+  public int getFlags() {
+    return (header[6] >> 5)& 0x7;
+  }
+
+  public int getFragment0() {
+    return (header[6] & 0x1f) << 8 + header[7];
+  }
+
+  public int getTTL() {
+    return header[8] & 0xff;
+  }
+
+  public int getProtocol() {
+    return header[9] & 0xff;
+  }
+
+  public int getChecksum() {
+    return get16(10);
+  }
+
+  public long getSourceIP() {
+    return get32(12);
+  }
+
+  public long getDestinationIP() {
+    return get32(16);
+  }
+
+  int get16(int pos) {
+    return ((header[pos] & 0xff) << 8) | (header[pos + 1] & 0xff);
+  }
+
+  long get32(int pos) {
+    return (((long) get16(pos)) << 16L) + get16(pos + 2);
+  }
+
+  int getData8(int pos) {
+    return data[pos] & 0xff;
+  }
+
+  int getData16(int pos) {
+    return ((data[pos] & 0xff) << 8) | (data[pos + 1] & 0xff);
+  }
+
+  long getData32(int pos) {
+    return (((long) get16(pos)) << 16L) + get16(pos + 2);
+  }
+
+  public static String getIPStr(long adr) {
+    return "" + ((adr >> 24) & 0xff) + "." + ((adr >> 16) & 0xff) + "." +
+      ((adr >> 8) & 0xff) + "." + (adr & 0xff);
+  }
+
+  public static void main(String[] args) throws IOException {
+    FileInputStream fs = new FileInputStream("test.txt");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fs*. Execution paths that do not contain closure statements were detected. Ensure *fs* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).